### PR TITLE
New save load interface

### PR DESCRIPTION
**Pull request by phlrain**
新增save load接口
静态图新增 
paddle.save( program, path )
paddle.load( program, path )

动态图新增

paddle.save_dygraph( state_dict,  path )
paddle.load_dygraph( path )

### CODE REVIEW

File: python/paddle/fluid/dygraph/layers.py [hunk 1]
```diff
@@ -24,8 +24,14 @@
 from .layer_object_helper import LayerObjectHelper
 from paddle.fluid import framework
 from ..param_attr import ParamAttr
+from paddle.fluid.framework import Variable
 
-__all__ = ['Layer']
+__all__ = ['Layer', 'var_base_to_np']
+
+
+def var_base_to_np(var_base):
```
Comment (JiabinYang) [inline python/paddle/fluid/dygraph/layers.py]: why not just use `numpy()`

File: python/paddle/fluid/dygraph/layers.py [hunk 2]
```diff
@@ -198,11 +204,33 @@ def add_parameter(self, name, parameter):
         """
         assert isinstance(parameter, framework.Parameter)
 
-        if parameter.name in self._loaddict_holder:
+        if len(self._loaddict_holder) > 0:
+            assert parameter.name in self._loaddict_holder, "Parameter not found, Can't not find [ {} ] in stat_dict".format(
+                parameter.name)
+
             var = parameter._ivar.value()
```
Comment (JiabinYang) [inline python/paddle/fluid/dygraph/layers.py]: maybe use `numpy()` is much easier?

File: python/paddle/fluid/dygraph/layers.py [hunk 3]
```diff
@@ -198,11 +204,33 @@ def add_parameter(self, name, parameter):
...
             tensor = var.get_tensor()
-            tensor.set(self._loaddict_holder[parameter.name].numpy(),
-                       framework._current_expected_place())
+
+            model_np = np.array(tensor)
+            load_para = self._loaddict_holder[parameter.name]
+
+            if isinstance(load_para, Variable):
```
Comment (JiabinYang) [inline python/paddle/fluid/dygraph/layers.py]: if use `numpy()` maybe we don't need to distinguish Variable and VarBase here?

File: python/paddle/fluid/dygraph/layers.py [hunk 4]
```diff
@@ -223,11 +251,35 @@ def __setattr__(self, name, value):
...
+                    value.name)
+
+                # check shape and dtype
                 var = value._ivar.value()
                 tensor = var.get_tensor()
-                tensor.set(self._loaddict_holder[value.name].numpy(),
-                           framework._current_expected_place())
+
+                model_np = np.array(tensor)
+                load_para = self._loaddict_holder[value.name]
```
Comment (JiabinYang) [inline python/paddle/fluid/dygraph/layers.py]: too many duplicated code here, make them in one function

File: python/paddle/fluid/io.py [hunk 5]
```diff
@@ -94,6 +95,10 @@ def is_persistable(var):
     return var.persistable
 
 
+def is_belong_to_optimizer(var):
```
Comment (JiabinYang) [inline python/paddle/fluid/io.py:98]: why we need this?

File: python/paddle/fluid/optimizer.py [hunk 6]
```diff
@@ -96,6 +98,78 @@ def __init__(self, learning_rate, regularization=None, name=None):
...
+                                          "Parameter shape not match, Dygraph Parameter [ {} ] need tensor with shape {} but load tensor with shape {}".format(
+                                                 item.name, model_np.shape, load_para_np.shape)
+
+                assert model_np.dtype == load_para_np.dtype, \
+                                          "Parameter dtype not match, Dygraph Parameter [ {} ] need tensor with dtype {}  but load tensor with dtype {}".format(
+                                                item.name, model_np.dtype, load_para_np.dtype)
+
+                tensor.set(load_para_np, framework._current_expected_place())
+
     def load(self, stat_dict):
```
Comment (JiabinYang) [inline python/paddle/fluid/optimizer.py]: remove this

File: paddle/fluid/framework/save_load_util.cc [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+    auto* data_ptr = tensor->data<void>();
+    if (platform::is_gpu_place(tensor->place())) {
+#ifdef PADDLE_WITH_CUDA
+      framework::Tensor temp;
+      TensorCopySync(*tensor, platform::CPUPlace(), &temp);
+      data_ptr = temp.data<void>();
+#else
+      PADDLE_THROW(
+          "Tensor is in CUDA device, but paddle not compile with CUDA, this "
+          "shoul'd not happen");
```
Comment (JiabinYang) [inline paddle/fluid/framework/save_load_util.cc]: should not

File: paddle/fluid/framework/save_load_util.cc [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+  }
+
+  ReadReserveBuffer(fin);
+
+  size_t tensor_number = ReadTensorNumber(fin);
+
+  for (size_t i = 0; i < tensor_number; ++i) {
+    std::string str_tensor_name = ReadTensorName(fin);
+
+    std::shared_ptr<Tensor> tensor_temp(new Tensor());
```
Comment (JiabinYang) [inline paddle/fluid/framework/save_load_util.cc:330]: why shared_ptr here?

File: python/paddle/fluid/dygraph/checkpoint.py [hunk 9]
```diff
@@ -16,12 +16,155 @@
...
-from ..framework import Variable, default_main_program
+from ..framework import Variable, default_main_program, in_dygraph_mode, dygraph_only
 import pickle
 from . import learning_rate_scheduler
 import warnings
+from .. import core
 
-__all__ = ['save_persistables', 'load_persistables']
+__all__ = [
+    'save_persistables', 'load_persistables', 'save_parameter',
```
Comment (JiabinYang) [inline python/paddle/fluid/dygraph/checkpoint.py]: remove this since it has problem

File: paddle/fluid/framework/save_load_util.cc [hunk 10]
```diff
@@ -0,0 +1,372 @@
...
+  }
+
+  fout.close();
+
+  return true;
+}
+
+bool LoadTensorFromDisk(
+    const std::string& file_name,
+    std::map<std::string, std::shared_ptr<Tensor>>& map_tensor) {  // NOLINT
```
Comment (JiabinYang) [inline paddle/fluid/framework/save_load_util.cc]: maybe use pointer instead of reference is better

File: python/paddle/fluid/dygraph/checkpoint.py (hunk 9 above)
Comment (JiabinYang) [inline python/paddle/fluid/dygraph/checkpoint.py]: and better not call it save_parameter, since fluid.save_parameter may cause confusion

**Review by JiabinYang — COMMENTED** [review]
some comments

File: python/paddle/fluid/dygraph/checkpoint.py (hunk 9 above)
Comment (JiabinYang) [inline python/paddle/fluid/dygraph/checkpoint.py]: and it would be better to make save as one API

File: paddle/fluid/framework/save_load_util.cc (hunk 10 above)
Comment (phlrain) [reply paddle/fluid/framework/save_load_util.cc]: result need to return

File: python/paddle/fluid/dygraph/checkpoint.py (hunk 9 above)
Comment (phlrain) [reply python/paddle/fluid/dygraph/checkpoint.py]: removed

File: paddle/fluid/framework/save_load_util.cc (hunk 8 above)
Comment (phlrain) [reply paddle/fluid/framework/save_load_util.cc:330]: tensor_temp  to put into map

File: paddle/fluid/framework/save_load_util.cc (hunk 7 above)
Comment (phlrain) [reply paddle/fluid/framework/save_load_util.cc]: done

File: python/paddle/fluid/optimizer.py (hunk 6 above)
Comment (phlrain) [reply python/paddle/fluid/optimizer.py]: done

File: python/paddle/fluid/io.py (hunk 5 above)
Comment (phlrain) [reply python/paddle/fluid/io.py:98]: filter out variable belong to optimizer

File: python/paddle/fluid/dygraph/layers.py (hunk 4 above)
Comment (phlrain) [reply python/paddle/fluid/dygraph/layers.py]: done

File: python/paddle/fluid/dygraph/layers.py (hunk 1 above)
Comment (phlrain) [reply python/paddle/fluid/dygraph/layers.py]: get numpy from varbase

File: python/paddle/fluid/dygraph/layers.py (hunk 2 above)
Comment (phlrain) [reply python/paddle/fluid/dygraph/layers.py]: support numpy, VarBase, Variable

File: python/paddle/fluid/tests/unittests/test_imperative_save_load.py [hunk 11]
```diff
@@ -363,8 +363,8 @@ def testLoadAndSetVarBase(self):
             if isinstance(adam._learning_rate, LearningRateDecay):
                 adam._learning_rate.step_num = 0
 
-            load_dict = fluid.load_optimizer("./test_dy")
-            adam.set_dict(load_dict)
+            para_state_dict, opti_state_dict = fluid.load_dygraph("./test_dy")
```
Comment (JiabinYang) [inline python/paddle/fluid/tests/unittests/test_imperative_save_load.py:366]: how about add an test to test if the program can work properly while loaded optimizer is None

**Review by JiabinYang — COMMENTED** [review]
some comments

**Review by JiabinYang — DISMISSED** [review]
LGTM

File: paddle/fluid/framework/save_load_util.cc [hunk 12]
```diff
@@ -0,0 +1,374 @@
...
+
+bool LoadStaticNameListFromDisk(
+    const std::string& file_name,
+    const std::vector<std::string>& vec_tensor_name_list, const Scope& scope) {
+  std::map<std::string, std::shared_ptr<Tensor>> map_load_tensor;
+  LoadTensorFromDisk(file_name, &map_load_tensor);
+
+  for (size_t i = 0; i < vec_tensor_name_list.size(); ++i) {
+    auto it = map_load_tensor.find(vec_tensor_name_list[i]);
+    PADDLE_ENFORCE(it != map_load_tensor.end(),
```
Comment (JiabinYang) [inline paddle/fluid/framework/save_load_util.cc:188]: maybe change this later， `PADDLE_ENFORCE_EQ` is recommended

**Review by raindrops2sea — APPROVED** [review]
